### PR TITLE
Issue #14: Duplicates in metadata.yml

### LINKED ISSUE #14: Duplicates in metadata.yml
We don't throw an error if some idiot (me) puts duplicate lines in metadata.yml

### PR DESCRIPTION
**Pull request by niallcreech**
Resolves #14 

Added a simple metadata.yml parsing function to remove duplicate
dependency entries. This does nothing spectacular, it just ignores
any formula named the same as one we already have. It does
this simply by name, ignoring differences in organisation or
anything else. Keeping this simple since there is a change in
metadata.yml format possibly upcoming which may make this
redundanct

### CODE REVIEW

File: shaker/helpers.py [hunk 1]
```diff
@@ -97,3 +98,59 @@ def validate_github_access(response):
...
+    Args:
+        metadata (dictionary): Keyed salt formula dependency information
+    
+    Returns:
+        resolved_dependencies (dictionary): The original metadata stripped of duplicates
+            If the metadata could not be resolved then we return the original args version
+    """ 
+    # Only start to make alterations if we have a valid metadata format
+    # Otherwise ignore and just return the one we were passed as an argument
+    if metadata and (type(metadata) == type({})) and ("dependencies" in metadata):
```
Comment (ashb) [inline shaker/helpers.py]: If the metadata is not valid it would be much preferable for the parse function (so `parse_metadata` rather than here) to throw a nice exception rather than returning a string and ending up with an unexpected/hard to diagnose error elsewhere in the code.

File: tests/test_metadata_handling.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    entry--
+        - dummy
+    """
+    @classmethod
+    def setup_class(cls):
+        pass
+    
+    @classmethod
+    def teardown_class(cls):
+        pass
```
Comment (ashb) [inline tests/test_metadata_handling.py]: Minor thing: these methods are optional and empty so could we remove them both please?

File: tests/test_metadata_handling.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    
+    def test_resolve_metadata_duplicates(self):
+        """
+        Check if we successfully remove duplicates from a sample metadata
+        """
+        original_metadata = yaml.load(self._sample_metadata_duplicates)
+        expected_metadata = yaml.load(self._sample_metadata_no_duplicates)
+        resolved_metadata = helpers.resolve_metadata_duplicates(original_metadata)
```
Comment (ashb) [inline tests/test_metadata_handling.py]: `resolve_metadata_duplicates` expects a dict so rather than using YAML as your test cases you can have objects directly. Something like

```python
    _sample_metadata_no_duplicates = {
        "dependencies": [
                "git@github.com:test_organisation/test1-formula.git",
                "git@github.com:test_organisation/test2-formula.git",
                "git@github.com:test_organisation/test3-formula.git",
        ]
        "entry" [ "dummy" ]
    }
```

This is a bit more direct to what we are testing.

**Comment by niallcreech** [comment]
@ashb Should be fixed up now

**Comment by ashb** [comment]
@niallcreech There's this comment too: https://github.com/ministryofjustice/salt-shaker/pull/21/files#r26854276 - do you think it makes sense?

**Comment by niallcreech** [comment]
Sorry, I'm not ignoring, just not getting any notification for PR comments for some reason :( 
I'll look at the comment, it makes sense to me but if it doesn't make sense to someone not me then that isnt any good!

File: tests/test_metadata_handling.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                            % (expected_metadata_entry))
+
+    def test_resolve_metadata_duplicates_bad_metadata_object(self):
+        """
+        Check if resolve_metadata_duplicates can handle bad yaml
+        data without breaking anything
+        """
+        expected = self._sample_metadata_bad_yaml
+        result = helpers.resolve_metadata_duplicates(self._sample_metadata_bad_yaml)
+        self.assertEqual(result, expected, "")
```
Comment (ashb) [inline tests/test_metadata_handling.py]: I think this test case got messed up - it's not really testing anything anymore.

We want to use something like:

```python
with self.assertRaises(TypeError):
  helpers.resolve_metadata_duplicates("")
``

(I think that's the syntax) Certainly it don't think that `_sample_metadata_bad_yaml` is useful with the last change.

**Comment by ashb** [comment]
One last change, sorry. We should add versions into the test fixtures so it's clearer which way it is resolving the duplicate. I think it's the first one?

```
...
            "git@github.com:test_organisation/test1-formula.git == v1.0.1",
            "git@github.com:test_organisation/test1-formula.git == v1.0.2",
            "git@github.com:test_organisation/test2-formula.git",
            "git@github.com:test_organisation/test3-formula.git",
            "git@github.com:test_organisation/test3-formula.git"
        ],
        "entry": ["dummy"]
    }

    _sample_metadata_no_duplicates = {
        "dependencies": [
            "git@github.com:test_organisation/test1-formula.git == v1.0.1",
...
```

**Comment by niallcreech** [comment]
Added some versioning to the test metadata to emphasize which duplicate entries are dropped and which kept

File: tests/test_metadata_handling.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+from shaker import helpers
+from nose.tools import raises
+
+class TestMetadataHandling(unittest.TestCase):
+
+    # Sample metadata with duplicates
+    _sample_metadata_duplicates = {
+        "dependencies": [
+            "git@github.com:test_organisation/test1-formula.git==v1.0.1",
+            "git@github.com:test_organisation/test1-formula.git==v1.0.1",
```
Comment (ashb) [inline tests/test_metadata_handling.py]: These two should have a different version, shouldn't they? And they'd still be resolved....

Or is this not how it works?

@mattmb This issue (#14) was reported by you - can you remember if you'd put the exact same formula+version twice or did you have different versions in each line?

If that's the case then @niallcreech we might need to move this up a layer to around here https://github.com/ministryofjustice/salt-shaker/blob/handle-metadata-duplicates/shaker/resolve_deps.py#L145-L150 - might need to rethink how we do this :(

(Sorry this is turning into such a back and forth PR – I'm doing a terrible job of either reviewing everything in one go and of providing enough detail of the ticket in the first place.)

Comment (niallcreech) [inline tests/test_metadata_handling.py]: It should have had a different version, the effect is the same though, the first one is dropped (theres no higher version resolutions or anything fancy) The versions are just there to illustrate which is kept. We could add something more 'intelligent' later but this is just to avoid a mess of duplicates overwriting each other

Comment (mattmb) [inline tests/test_metadata_handling.py]: I have a feeling it was a duplicate with the same version and something nasty happens in the recursive logic just by processing it twice.

Comment (mattmb) [inline tests/test_metadata_handling.py]: Could be wrong though, I didn't really add enough data to that issue :-(

**Comment by niallcreech** [comment]
Fixed sample metadata typo

**Comment by ashb** [comment]
\o/ Sorry it took so long but we got there in the end

**Comment by niallcreech** [comment]
Like a fine wine, it cannot be rushed :)
